### DISCUSSION TITLE
Warehouse Robotics Simulation (Project 212)

### DISCUSSION THREAD

**Discussion by robertogl** [mathworks/MATLAB-Simulink-Challenge-Project-Hub#41] [Collaborate]
Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project 212](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Warehouse%20Robotics%20Simulation)

---
*:robot: An AI bot may reply to your comments with suggestions and guidance. It will not solve the project for you. Your comments may be processed by a third-party LLM. To direct a question to a specific person, tag them with @username.*

**Comment by braenmachio** [comment]
Hey, I'm Machio Brian [Kenya], I'll be working on Project 212

**Comment by adityabastapure** [comment]
In Project 212
How can we create different warehouse models?
I was able to create in 2D but I don't how to create a model in gazebo?
Also can I create dynamic obstacle in 2D environment ?
And how to implement dynamic obstacle avoidance in 2D environment

**Reply by robertogl** [reply]
Hi @adityabastapure,

We have several examples with a 3D warehouse-like environment created in Gazebo. This [example](https://www.mathworks.com/help/robotics/ug/a-star-path-planning-and-obstacle-avoidance.html) is one of them and you can find the Gazebo scenario within the [Virtual Machine](https://www.mathworks.com/support/product/robotics/ros2-vm-installation-instructions.html) that we ship with the examples. You can start with the 'Gazebo Warehouse Robot with Obstacles' world and modify it as you need.

Regarding obstacle avoidance, we ship an implementation of the [Vector Field Histograms (VFH)](https://www.mathworks.com/help/nav/ug/vector-field-histograms.html) algorithm as part of the [Navigation Toolbox](https://www.mathworks.com/help/nav/index.html). The name of the function is [controllerVFH](https://www.mathworks.com/help/nav/ref/controllervfh-system-object.html).

I hope this helps!

**Comment by adityabastapure** [comment]
mobileRobotPRM difference plaannerPRM matlab
please reply as soon as possible

**Reply by robertogl** [reply]
Hi @adityabastapure,

The main difference is that PlannerPRM is customizable and more generic as you can add different statespaces.

**Reply by adityabastapure** [reply]
Thank you @robertogl for your quick support

**Reply by adityabastapure** [reply]
> mobileRobotPRM difference plaannerPRM matlab please reply as soon as possible

Case I :
image = imread('C:\Users\aabas\OneDrive\Desktop\trialblock2.png');
grayimage = rgb2gray(image);
bwimage = grayimage < 0.5;
map = binaryOccupancyMap(bwimage,20);

prm = mobileRobotPRM;
prm.Map = map;
prm.NumNodes = 100;
prm.ConnectionDistance = 5;

Case II :
Same how can implement above using plannerPRM or planner = plannerRRTStar(stateSpace,stateVal) or plannerAStarGrid  how to define parametes here?
how to find path on generated map? initially for first case i used [xy]= findpath(prm,start,goal)

please reply as soon as possible

**Reply by adityabastapure** [reply]
load exampleMaps
map = binaryOccupancyMap(simpleMap);

start = [5,5];
goal = [20,20];

planner = plannerAStarGrid(map);
path = plan(planner,start,goal)

subplot(1,2,1)
show(map)
hold on

getting error while plotting this

**Reply by adityabastapure** [reply]
![Capture](https://user-images.githubusercontent.com/19888857/173644976-cc0294e5-2b2c-4476-a688-30ac549a329c.PNG)

**Reply by adityabastapure** [reply]
ss = stateSpaceSE2;
sv = validatorOccupancyMap(ss);
load exampleMaps
map = occupancyMap(simpleMap);
sv.Map = map;
sv.ValidationDistance = 0.01;

planner = plannerRRT(ss,sv);

planner.MaxConnectionDistance = 2.5;
% planner.BallRadiusConstant = 5;
planner.MaxNumTreeNodes = 100;
start = [5,5,0];
goal = [20,20,0];

[pthObj,solnInfo] = plan(planner,start,goal)

subplot(1,2,1)
show(map)
hold on

plot(solnInfo.TreeData(:,1),solnInfo.TreeData(:,2),'.-'); % tree expansion
plot(pthObj.States(:,1),pthObj.States(:,2),'r-','LineWidth',2) % draw path

path =  pthObj.States(:,1:2)

controller = controllerPurePursuit;
controller.Waypoints = path;
controller.DesiredLinearVelocity = 0.6;
controller.MaxAngularVelocity = 2;
controller.LookaheadDistance = 1;
![Capture2](https://user-images.githubusercontent.com/19888857/173648393-7ace0960-3c3f-4352-9b52-7294115752e9.PNG)

**Reply by venkateshbs** [reply]
> ![Capture](https://user-images.githubusercontent.com/19888857/173644976-cc0294e5-2b2c-4476-a688-30ac549a329c.PNG)

I will get back to you on this one soon.

**Reply by ankur-sriv** [reply]
> load exampleMaps map = binaryOccupancyMap(simpleMap);
> 
> start = [5,5]; goal = [20,20];
> 
> planner = plannerAStarGrid(map); 
> path = plan(planner,start,goal)
> 
> subplot(1,2,1) show(map) hold on
> 
> getting error while plotting this.


Hi @adityabastapure,

`plannerAStarGrid` works by default for grids, hence start & goal given currently are considered as grid points whose origin is top-left corner. Hence, plot seems wrong.

I understand that you want to search in the world frame of reference with 'X & Y'.
For that you must call `plan(planner,start,goal,'world')` instead of `plan(planner,start,goal)` as given below. 

> load exampleMaps;
> map = binaryOccupancyMap(simpleMap);
> 
> start = [5,5]; 
> goal = [20,20];
> 
> planner = plannerAStarGrid(map); 
> path = plan(planner,start,goal. 'world') 
> show(planner);
>

**Comment by adityabastapure** [comment]
ss = stateSpaceSE2;
sv = validatorOccupancyMap(ss);
load exampleMaps
map = occupancyMap(simpleMap);
sv.Map = map;
sv.ValidationDistance = 0.01;
planner = plannerRRTStar(ss,sv);

% planner.MaxConnectionDistance = 2.5;
planner.BallRadiusConstant = 10;   ***************************** This parameter is not showing any effect on output
% planner.MaxNumTreeNodes = 100;

start = [5,5,0];
goal = [20,20,0];

[pthObj,solnInfo] = plan(planner,start,goal)

subplot(1,2,1)
show(map)
hold on

plot(solnInfo.TreeData(:,1),solnInfo.TreeData(:,2),'.-'); % tree expansion
plot(pthObj.States(:,1),pthObj.States(:,2),'r-','LineWidth',2) % draw path

path =  pthObj.States(:,1:2)

% Calculates Path Length
j = size(path);
j = j(1)
Path_Length = 0;
i = 1;
while i < j
    W1 = path(i,:);
    W2 = path(i+1,:);
    distance = norm(W1 - W2)
    Path_Length = Path_Length + distance;
    i = i + 1;
end
Path_Length


RRT* should have more optimal solution than RRT
depending upon planner.BallRadiusConstant = 10;   ***************************** This parameter is not showing any effect on output

But its exactly reverse RRT is having optimal solution than RRT*
Please help me out fast

**Reply by ajay102p** [reply]
Could you clarify on what you expect on RRTStar output compared to RRT. Is it path length you are referring to?
You could try the [plannerBenchmark ](https://www.mathworks.com/help/nav/ref/plannerbenchmark.html) feature to find which planner provide the least path length . Also for computing path length, consider creating [navPath](https://www.mathworks.com/help/nav/ref/navpath.html?s_tid=doc_ta) object, which can easily calculate path length.

**Comment by adityabastapure** [comment]
for PlannerRRT/PlannerRRTStar
IsPathFound: 0

Is their any threshold radius around goal point to consider sampled node is equivalent to goal point

If algorithm compares sampled node exactly to the goal coordinate the solution will be not found easily

**Reply by adityabastapure** [reply]
I need to run planner until and unless solution is found

**Reply by ajay102p** [reply]
Hi @adityabastapure,

The default GoalReachedFcn is nav.algs.checkIfGoalIsReached, which has a tolerance of 0.5 meters. That means the planner will be able to find a path even if the  goal point is in 0.5m radius. You could easily customize this function by editing the default function and provide the function handle to the GoalReachedFcn property of [plannerRRTStar](https://www.mathworks.com/help/nav/ref/plannerrrtstar.html).

**Comment by ozayp** [comment]
Hi am ozayo from wits university (South Africa), am interested in robotics and am currently learning BSc in mechanical engineering. I am looking for collaboration in project title: Warehouse Robotics Simulation.

**Comment by thokozanimnguni** [comment]
hi everyone im Thokozani from wits university (South Africa) and studying towards a (BSc Mechanical engineering). i am currently looking for collaborators in the Warehouse Robotics Simulation project

**Comment by Prajwal51122** [comment]
Hii everyone, I am Prajwal Ahire  from Walchand College of Engineering ( India ). I aminterested in working and collaborating on this project

**Comment by HLLMIC031** [comment]
Hello, I'm Emilia Skye Hills from the University of Cape Town working on this project. I am also wondering if anyone else got stuck waiting to download the VMWare Workstation due to "Account Verification Pending"?

**Comment by mechdrive** [comment]
Hello everyone,
I see the official MathWorks VM uses Ubuntu 18.04 with ROS Melodic/Dashing and Gazebo 9 for the challenge examples.
Can I use a more recent setup, like Ubuntu 22.04 and ROS 2 Humble, for the Warehouse Robotics Simulation project? I am using MATLAB R2024b which supports ROS 2 Humble.
Will I face any integration issues with Gazebo or the example models, or are there recommended adaptations for running this project on the newer systems?

**Reply by robertogl** [reply]
Hi @mechdrive,

Sorry for the late response. It should work but we have not tested it yet, so I cannot confirm. The latest is Ubuntu 20.04 with ROS 2 Humble as you can see [here](https://www.mathworks.com/support/product/robotics/ros2-vm-installation-instructions-v9.html).
